sdk/identity/azure-identity/azure/identity/_credentials/device_code.py

```diff
@@ -61,6 +66,11 @@ def __init__(self, client_id, **kwargs):
         self._prompt_callback = kwargs.pop("prompt_callback", None)
         super(DeviceCodeCredential, self).__init__(client_id=client_id, **kwargs)
 
+    @log_get_token(_LOGGER, "DeviceCodeCredential")
```

Comment:
Does self.__class__.__name__ work here? Then we don't need to have this in the sub-classes.

Reply:
I wish it did. However, the decorator is applied during parsing of the class definition. There's no `self` and the class isn't yet defined (so this can't be just `DeviceCodeCredential`).

Reply:
Then I doubt if decorator is the right way to go. 

It will require us dup a lot of code for some complex hierarchy。

Reply:
I'm interested in better ways to do this but I don't see duplication as one of the problems here. The decorator adds 2 lines of boilerplate (import the decorator, apply it to `get_token`) to every credential module. How could we get the same result with less duplication? What do you mean by "complex hierarchy"?

Reply:
I meant if the hierarchy is very deep and wide like a tree, we would not be able to fully benefit from polymorphism.

And my another concern is like
class A()
class B(A)
class C(B)

My expected behavior: when I call A->get_token(), it can recognize the real class is C hence it calls C->get_token() and only logs C->get_token() is called.

In our case, C->get_token() ("C is called is logged") will call into B->get_token() and call into A->get_token(), will we have "B is called" & "A is called" in our log?

Reply:
First, let's agree not to build a deep and wide inheritance hierarchy 🤝 

We're talking about an instance method. Calling `C().get_token()` only invokes `A().get_token()` if it calls `super` or `C` doesn't override `get_token`. But yes, if one decorated method calls another, each adds log entries. This isn't particular to the decorator though. Imagine the obvious alternative implementation: every `get_token` does its own logging. Then if `C.get_token()` calls `A().get_token()`, you still get logs from each. Nor is this necessarily a problem. One of our goals is to provide insight into what composite credentials like `DefaultAzureCredential` are doing. We want to see logs from both `DefaultAzureCredential` and the inner credentials it invokes.

Reply:
I think we already have it? 

UsernamePasswordCredential, InteractiveBrowserCredential, DeviceCodeCredential all inherit from InteractiveCredential which is a subclass of PublicClientCredential which is a sub class of MsalCredential.

Though now we are not quite hit by the problem because get_token in PublicClientCredential & MsalCredential are abstractmethod. We need to think about the cost of one day it happens.

Another question is let's say we want the logs of UsernamePasswordCredential, InteractiveBrowserCredential & DeviceCodeCredential  to be a little bit different. How can we achieve it?

Reply:
> Another question is let's say we want the logs of UsernamePasswordCredential, InteractiveBrowserCredential & DeviceCodeCredential to be a little bit different. How can we achieve it?

The differences in their behavior are in their `_request_token` methods. We can log anything specific to an implementation in that method.